extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,287 @@
+      templateData: null,
+      crossDomainData: {
+        rawCreativeBytes: this.unvalidatedBytes_,
+        additionalContextMetadata: this.getCrossDomainContextData(),
+        sentinel: this.getCrossDomainSentinel(),
+      },
+      unvalidatedBytes: this.unvalidatedBytes_,
+      // TODO(levitzky) This may change based on the ad response.
+      size: this.initialSize_,
+      adUrl: this.expandedAdUrl_ ? this.expandedAdUrl_ : this.adUrl_,
+    });
+  }
+
+  /**
+   * Processes the ad response as soon as the XHR request returns. This can be
+   * overridden and used as a hook to perform any desired logic before passing
+   * the response to the validator.
+   * @param {{bytes: !ArrayBuffer, headers: !Headers}} response
+   * @protected
+   */
+  handleAdResponse(response) {
+    if (this.isStale_()) {
+      return;
+    }
+    const unvalidatedBytes = response.bytes;
+    const headers = response.headers;
+    if (!unvalidatedBytes) {
+      // TODO(levitzky) Add error reporting.
+      this.forceCollapse();
+      return;
+    }
+    dev().assert(this.boundValidator_, 'Validator never bound!');
+    this.unvalidatedBytes_ = unvalidatedBytes;
+    this.boundValidator_(unvalidatedBytes, headers, this)
+        .then(validatedBytes =>
+          this.handleValidatorResponse(validatedBytes));
+  }
+
+  handleAdResponseError(error) {
+    // TODO(levitzky) add actual error processing logic.
+    dev().warn(TAG, error);
+  }
+
+  /**
+   * Handler invoked when the current freshness id has expired.
+   */
+  handleStaleExecution() {}
+
+  /**
+   * Processes validator response and delegates further action to appropriate
+   *   renderer.
+   * @param {!ValidatorOutputDef} validatedResponse The utf-8 decoded ad
+   *   response.
+   */
+  handleValidatorResponse(validatedResponse) {
+    if (this.isStale_()) {
+      return;
+    }
+    if (!validatedResponse.creative) {
+      // TODO(levitzky) Add error reporting.
+      this.forceCollapse();
+      return;
+    }
+    dev().assert(this.boundRenderers_[validatedResponse.result],
+        'Renderer for AMP creatives never bound!');
+    this.boundRenderers_[validatedResponse.result](
+        this.getRenderingDataInput_(validatedResponse), this);
+  }
+
+
+  resetInstance() {
+    this.freshnessId_++;
+    this.uiHandler.applyUnlayoutUI();
+  }
+
+  /** @override */
+  buildCallback() {
+    this.uiHandler = new AMP.AmpAdUIHandler(this);
+  }
+
+  /** @override */
+  isLayoutSupported(unusedLayout) {
+    return true;
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    ++this.freshnessId_;
+    this.isStale_ = this.getFreshnessVerifier();
+    sendXhrRequest(this.getExpandedUrl_())
+        .then(response => this.handleAdResponse(response))
+        .catch(error => this.handleAdResponseError(error));
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    this.freshnessId_++;
+    return true;
+  }
+}
+
+AMP.extension(TAG, '0.1', AMP => {
```

Comment:
this should not be an extension, it's a base class

Reply:
Done.